### PR TITLE
enhance flatten_datacell functions

### PR DESCRIPTION
**Pull request by LHT129**
- add serialize and deserialize
- add basic test for flatten interface

### CODE REVIEW

File: src/data_cell/flatten_interface.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    std::string io_string = json_obj[IO_TYPE_KEY];
+    if (io_string == IO_TYPE_VALUE_BLOCK_MEMORY_IO) {
+        return make_instance<MemoryBlockIO>(json_obj, common_param);
+    }
+    if (io_string == IO_TYPE_VALUE_MEMORY_IO) {
+        return make_instance<MemoryIO>(json_obj, common_param);
+    }
+    return nullptr;
+}
+}  // namespace vsag
```
Comment (wxyucs) [inline src/data_cell/flatten_interface.cpp]: add a blank line at the end of file

File: src/data_cell/flatten_interface.h [hunk 2]
```diff
@@ -33,38 +46,57 @@ class FlattenInterface {
...
+
 public:
-    uint64_t total_count_{0};
-    uint64_t max_capacity_{1000000};
-    uint64_t code_size_{0};
+    InnerIdType total_count_{0};
+    InnerIdType max_capacity_{1000000};
+    uint32_t code_size_{0};
 };
+
```
Comment (wxyucs) [inline src/data_cell/flatten_interface.h:104]: please add an alias here, `using FlattenInterfacePtr = std::shared_ptr<FlattenInterface>`, and use it other where

File: src/data_cell/flatten_interface_test.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    TestSerializeAndDeserialize(int64_t dim,
+                                std::shared_ptr<FlattenInterface> other,
+                                float error = 1e-5f);
+
+public:
+    std::shared_ptr<FlattenInterface> flatten_{nullptr};
+
+    MetricType metric_{MetricType::METRIC_TYPE_L2SQR};
+};
+}  // namespace vsag
```
Comment (wxyucs) [inline src/data_cell/flatten_interface_test.h]: add a blank line at the end of file

**Review by wxyucs — CHANGES_REQUESTED** [review]

File: src/data_cell/flatten_interface.h (hunk 2 above)
Comment (LHT129) [reply src/data_cell/flatten_interface.h:104]: done

File: src/data_cell/flatten_interface_test.h (hunk 3 above)
Comment (LHT129) [reply src/data_cell/flatten_interface_test.h]: done

File: src/data_cell/flatten_interface.cpp (hunk 1 above)
Comment (LHT129) [reply src/data_cell/flatten_interface.cpp]: done

File: src/data_cell/flatten_datacell.h [hunk 4]
```diff
@@ -25,16 +25,21 @@ namespace vsag {
 
 template <typename QuantTmpl, typename IOTmpl>
 class FlattenDataCell : public FlattenInterface {
+    /*
```
Comment (jiaweizone) [inline src/data_cell/flatten_datacell.h]: Move to L25 ?

Comment (LHT129) [reply src/data_cell/flatten_datacell.h]: done

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by wxyucs — APPROVED** [review]
lgtm

**Pull request merged**